### PR TITLE
Register annotated commands in Drupal Console.

### PR DESCRIPTION
**Pull request by greg-1-anderson**
This PR adds annotation command support to DrupalConsole.

Any command file `FooCommand.php` may now contain either a DrupalConsole Symfony Console Command implementation, in which case it works as it always has, or, alternately, if the class FooCommand does not extend Drupal\Console\Command\Command, then it is presumed that the instance is an annotation command, and all of its public methods are registered as DrupalConsole commands.

The purpose of this is to allow Drupal modules to implement simple, loosely-coupled commandline tool interfaces. These commands might be implemented by extending [Robo](https://github.com/Codegyre/Robo), although this is not required.

I was planning on converting the DrupalConsole InitCommand into an annotation class, simply to demonstrate how that is done; however, since this command is already tightly coupled with Drupal\Console\Command\Command helpers, `getModule()`, etc., doing this would be counterproductive; in general, it is a non-goal to convert core DrupalConsole commands into annotation commands. For a typical simple Drupal module, though, whose main purpose is to provide a CLI interface that simply calls a method or two in the module, the annotation class will be a win.

Follow development in Drush at https://github.com/drush-ops/drush/issues/2103.

DrupalConsole checklist (to consider / discuss):

- [x] Allow Command object to be passed in as a parameter to task methods that need it. https://github.com/consolidation-org/annotation-command/pull/4
- [x] Allow $input and $output objects to be passed in as parameters. https://github.com/consolidation-org/annotation-command/pull/5
- [ ] Use league/container so that we have use of inflectors https://github.com/hechoendrupal/DrupalConsole/issues/2061
- [ ] Use a styled PSR-3 logger https://github.com/consolidation-org/log
- [ ] Consider including Robo in composer.json, so Annotated commands have access to tasks.
- [ ] Translation strategy
  - [ ] Annotation strings
  - [ ] PSR-3 logger strings

### CODE REVIEW

**Comment by greg-1-anderson** [comment]
Replaced old technique with something better. No longer overloads the "*Command.php" namespace, and instead looks for command files following their own rules.

Untested, but on the right track.

**Comment by greg-1-anderson** [comment]
The latest commit (12daaa6) works great with the [sample cli command I wrote (ported) for the default_content module](https://www.drupal.org/node/2703131). Search paths and naming conventions still need to be discussed and finalized. I am not set on (or even very happy with) the placeholder "CliTools" folder name that I am using now; just picked something quickly to get started.

Does not work with Drush yet, but it will only be a small effort to fix up the PR there as well; I'm not going to do that tonight, but will have it working in short order.

**Comment by greg-1-anderson** [comment]
Now also works with Drush.

**Comment by greg-1-anderson** [comment]
Discussion on search paths and naming conventions for command files at https://github.com/consolidation-org/annotation-command/issues/12

**Comment by greg-1-anderson** [comment]
Just for fun, I added a generate:meme example command. We can take this out prior to merge if desired; in the interim, it is useful to have a test command bundled with the PR.

Uses Impact font on a Mac. If running on some other platform, download Anton.ttf from https://googlefontdirectory.googlecode.com/hg/ofl/anton/ and place Anton.ttf in ~/.fonts.

**Comment by jmolivas** [comment]
@greg-1-anderson  Nice usage of annotations, we have been discussing using them before as you can see on issue [Use annotations in Commands](https://github.com/hechoendrupal/DrupalConsole/issues/478) and the related PR [Annotations in Drupal Commands](https://github.com/hechoendrupal/DrupalConsole/pull/479).

Our proposal for using annotation was for other purpose. We wanted annotations to define any behavior not related to Symfony/Console Command and make our commands as much as close to the official standard. [Creating a basic Command - Symfony documentation](http://symfony.com/doc/current/components/console/introduction.html#creating-a-basic-command)

#### As stated on issue definition
> The Drupal commands interact with more stuffs that the commands in Symfony, in Drupal, one command can be have many dependencies with other modules or themes or any other scope in your installation.

> This approach is to use Annotations to provide metadata on each command to specify dependencies or check another variables in Drupal or any other stuff that interacts with the console.

> For example:
```php
/**
 * @DrupalCommand (
 *      dependencies = {
 *          "simpletest"
 *      },
 * )
 */
class TestDebugCommand extends ContainerAwareCommand { ... }
```

With that said, after discussing with the other project maintainers we don’t agree with creating a new Drupalisim or a new Drupal way to write Symfony/Console based Commands when there is a well accepted and defined standard provided and supported by the community.

We really like the idea about decouple code create components to reuse between different Drupal CLI tools.

**Comment by greg-1-anderson** [comment]
Well, this is very disappointing, and more than a little ironic; however, I think there is nothing to do here at this point but close the PR and move on.

**Comment by greg-1-anderson** [comment]
I have an idea on how we could configure Symfony/Console commands without diverging from the Symfony API.  Re-opening so I can push another commit here to demonstrate.

File: src/Command/MemeCommand.php [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+class MemeCommand extends Command
```
Comment (greg-1-anderson) [inline src/Command/MemeCommand.php]: Commands extend Drupal\Console\Command\Command, just as before. These are Symfony commands.

File: src/Command/MemeCommand.php [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+     * @command generate:meme
+     * @param string $topMessage The message to place at the top of the image.
+     * @param string $bottomMessage The message to place at the bottom of the image.
+     * @param string $image The path to the meme image to use.
+     * @default $image meme.png
+     * @option string $out The file to write the generated meme to.
+     * @default $out generated-meme.png
+     * @usage generate:meme "I don't always make memes" "But when I do, I use the command line" my-source-image.png
+     */
+    public function execute(InputInterface $input, OutputInterface $output)
```
Comment (greg-1-anderson) [inline src/Command/MemeCommand.php]: The configure() method is replaced with annotations on the execute() method. These annotations could also go on the class docblock; I'm not sure which is better.

File: src/Command/Settings/InitCommand.php [hunk 3]
```diff
@@ -19,24 +19,13 @@
...
-     * {@inheritdoc}
+     * Init command
+     *
+     * Copy configuration files to user home directory.
+     *
+     * @command settings:init
+     * @option boolean $override Override configurations files
+     * @aliases init
      */
     protected function execute(InputInterface $input, OutputInterface $output)
```
Comment (greg-1-anderson) [inline src/Command/Settings/InitCommand.php]: Again, porting existing Drupal Console commands to use annotations is not required; however, I did convert the settings:init command simply to allow configuration via annotations to be compared side-by-side with configuration via php method calls.

File: src/Command/Command.php [hunk 4]
```diff
@@ -2,15 +2,15 @@
...
+use Consolidation\AnnotatedCommand\AnnotatedCommand;
 use Symfony\Component\Console\Helper\HelperSet;
 use Drupal\Console\Helper\HelperTrait;
 
 /**
  * Class Command
  * @package Drupal\Console\Command
  */
-abstract class Command extends BaseCommand
+abstract class Command extends AnnotatedCommand
```
Comment (greg-1-anderson) [inline src/Command/Command.php]: Drupal\Console\Command now extends from AnnotatedCommand.  AnnotatedCommand extends Symfony\Component\Console\Command\Command, and will behave exactly like a Symfony/Console Command unless there are annotations present that configure the command.  Even when annotations are used, the base class will still operate the same way, calling `execute` with `$input` and `$output` variables when the command is used.

**Comment by greg-1-anderson** [comment]
As you can see from the diffs in this PR, the code now conforms to the Symfony/Console command API exactly, with annotations being used only as an optional configuration method. Having this in place also gets us one step closer to being able to use annotations to specify dependencies, as referenced in the comments above.

The importance of this PR extends beyond the simple application of annotations for configuration. Using a common base class from Consolidation-org will allow us to do things such as share the [output formatters](https://github.com/consolidation-org/output-formatters) library between Drush and Drupal Console. Drupal Console has a similar feature, but the consolidation output formatters are more complete, and using a common implementation will allow us to present a uniform interface for Drupal modules to use.

I hope that you will find this new implementation more palatable than the previous one. Using annotations in Drupal Console core is not required; accepting this PR would allow module developers to use or not use them as they saw fit.

**Comment by jmolivas** [comment]
Looking at this, thanks for the update @greg-1-anderson

**Comment by jmolivas** [comment]
Support for annotations merged on PR https://github.com/hechoendrupal/DrupalConsole/pull/2642

**Pull request closed without merging**